query/control/controller.go

```diff
@@ -512,15 +512,24 @@ func (c *Controller) Queries() []*Query {
 // This will return once the Controller's run loop has been exited and all
 // queries have been finished or until the Context has been canceled.
 func (c *Controller) Shutdown(ctx context.Context) error {
+	// Wait for query processing goroutines to finish.
+	defer c.wg.Wait()
+
 	// Mark that the controller is shutdown so it does not
 	// accept new queries.
-	c.queriesMu.Lock()
-	c.shutdown = true
-	if len(c.queries) == 0 {
-		c.queriesMu.Unlock()
-		return nil
```

Comment:
This was racy - even if all the queries are removed from the map, that doesn't mean that all the spawned goroutines are finished. When I added the zaptest logging I started seeing failures from the race detector because we would get a bit of logging after the end of the test, when the test logger was no longer valid.

Reply:
Is the same problem in 2.x?

Reply:
Likely yes.